### ISSUE TITLE
HTML5 parser closes <math> before <span>

### ISSUE THREAD

**Issue by syakovyn** [sparklemotion/nokogiri#2335]
HTML5 parser differs in output from HTML4 parser. We consider the HTML4 parser output is the right one.

A script to reproduce the bug:

```ruby
#! /usr/bin/env ruby

require 'nokogiri'
require 'minitest/autorun'

class Test < MiniTest::Spec
  describe "HTML5 parser" do
    it "produces the same output as HTML4" do
      html = <<-HTML
<mjx-container>
<mjx-assistive-mml>
<math xmlns=\"http://www.w3.org/1998/Math/MathML\">
            <msqrt><span>\n\n  </span>
                <mn>3</mn>
            </msqrt>
        </math>
        <script>window.destroyAllHumans()</script>
        \n\n
        <msqrt>
            <mn>4</mn>
        </msqrt>
    </mjx-assistive-mml>
</mjx-container>
      HTML
      html5 = Nokogiri::HTML5.fragment(html).to_s
      html4 = Nokogiri::HTML4.fragment(html).to_s

      assert_equal html4, html5
    end
  end
end
```

**Environment**

```
# Nokogiri (1.12.5)
    ---
    warnings: []
    nokogiri:
      version: 1.12.5
      cppflags:
      - "-I/Users/syakovyn/.rvm/gems/ruby-2.7.4/gems/nokogiri-1.12.5-x86_64-darwin/ext/nokogiri"
      - "-I/Users/syakovyn/.rvm/gems/ruby-2.7.4/gems/nokogiri-1.12.5-x86_64-darwin/ext/nokogiri/include"
      - "-I/Users/syakovyn/.rvm/gems/ruby-2.7.4/gems/nokogiri-1.12.5-x86_64-darwin/ext/nokogiri/include/libxml2"
      ldflags: []
    ruby:
      version: 2.7.4
      platform: x86_64-darwin20
      gem_platform: x86_64-darwin-20
      description: ruby 2.7.4p191 (2021-07-07 revision a21a3b7d23) [x86_64-darwin20]
      engine: ruby
    libxml:
      source: packaged
      precompiled: true
      patches:
      - 0001-Remove-script-macro-support.patch
      - 0002-Update-entities-to-remove-handling-of-ssi.patch
      - 0003-libxml2.la-is-in-top_builddir.patch
      - 0004-use-glibc-strlen.patch
      - 0005-avoid-isnan-isinf.patch
      - 0006-update-automake-files-for-arm64.patch
      - 0007-Fix-XPath-recursion-limit.patch
      libxml2_path: "/Users/syakovyn/.rvm/gems/ruby-2.7.4/gems/nokogiri-1.12.5-x86_64-darwin/ext/nokogiri"
      memory_management: ruby
      iconv_enabled: true
      compiled: 2.9.12
      loaded: 2.9.12
    libxslt:
      source: packaged
      precompiled: true
      patches:
      - 0001-update-automake-files-for-arm64.patch
      - 0002-Fix-xml2-config-check-in-configure-script.patch
      datetime_enabled: true
      compiled: 1.1.34
      loaded: 1.1.34
    other_libraries:
      zlib: 1.2.11
      libiconv: '1.15'
      libgumbo: 1.0.0-nokogiri
```

**Comment by flavorjones** [comment]
@syakovyn Thanks for submitting this, and sorry you're having problems.

I've reproduced the problem, but have a slightly different interpretation of what's happening. Here's the script I'm using:

```ruby
#! /usr/bin/env ruby

require "nokogiri"

html = <<~EOF
  <math xmlns=\"http://www.w3.org/1998/Math/MathML\">
    <msqrt><mn>4</mn></msqrt>
  </math>
EOF

puts Nokogiri::HTML5.fragment(html).to_html
# => <math xmlns="http://www.w3.org/1998/Math/MathML">
#      <msqrt><mn>4</mn></msqrt>
#    </math>

html = <<~EOF
  <math xmlns=\"http://www.w3.org/1998/Math/MathML\">
    <msqrt><span></span><mn>3</mn></msqrt>
  </math>
EOF

puts Nokogiri::HTML5.fragment(html).to_html
# => <math xmlns="http://www.w3.org/1998/Math/MathML">
#      <msqrt></msqrt></math><span></span><mn>3</mn>
```

My guess at the root cause is that the `<span>` tag is forcing early closure of the `<math>` tag.

@stevecheckoway Can you validate my interpretation of what Gumbo is doing here? If you're able, I'd love to pair with you on debugging and fixing the behavior.

**Comment by flavorjones** [comment]
More information:

```ruby
html = <<~EOF
  <math xmlns=\"http://www.w3.org/1998/Math/MathML\">
    <msqrt><span></span><mn>3</mn></msqrt>
  </math>
EOF

pp Nokogiri::HTML5.fragment(html, max_errors: 100).errors
```

outputs

```text
[#<Nokogiri::XML::SyntaxError: 2:10: ERROR: That tag isn't allowed here  Currently open tags: html, math, .
  <msqrt><span></span><mn>3</mn></msqrt>
         ^>,
 #<Nokogiri::XML::SyntaxError: 2:33: ERROR: That tag isn't allowed here  Currently open tags: html.
  <msqrt><span></span><mn>3</mn></msqrt>
                                ^>,
 #<Nokogiri::XML::SyntaxError: 3:1: ERROR: That tag isn't allowed here  Currently open tags: html.
</math>
^>]
```

**Comment by flavorjones** [comment]
This looks to me like Gumbo considers a `<span>` element within a `<math>` block to be invalid. I'm not familiar with the MathML spec and so will defer to @syakovyn and @stevecheckoway and others to discuss.

**Comment by stevecheckoway** [comment]
I can track down the exact portion of the spec that mandates this behavior, but I think the easiest thing to do is to open a document in a browser and use the browser's web tools to inspect the DOM. Alternatively, you can use JavaScript to print the `body` elements' `innerHTML`. Let's do that:
```html
<!DOCTYPE html>
<math xmlns="http://www.w3.org/1998/Math/MathML">
  <msqrt><span></span><mn>4</mn></msqrt>
</math>
<script>
alert(document.body.innerHTML)
</script>
```

If you save that as a file and open it, it'll pop up an alert containing (and I've reformatted the HTML slightly)
```
<math xmlns="http://www.w3.org/1998/Math/MathML">
  <msqrt></msqrt>
</math>
<span></span>
<mn>4</mn>
```

In other words, I believe the HTML5 parser is correctly parsing this document.

**Comment by stevecheckoway** [comment]
And for what it's worth, this is the [section of the HTML standard](https://html.spec.whatwg.org/multipage/parsing.html#parsing-main-inforeign) that mandates this behavior. When the parser encounters the `span` start tag, it is in a foreign context (because the parent element is a `math` element in the MathML namespace). The rule
> A start tag whose tag name is one of: … "span" …
applies and that mandates a parse error and, in this case, closing the `msqrt` and `math` elements.

The other two errors @flavorjones's [showed](https://github.com/sparklemotion/nokogiri/issues/2335#issuecomment-932354599) come from the browser encountering `</msqrt>` and then `</math>` without there being open `msqrt` and `math` elements.

**Comment by syakovyn** [comment]
As far as I understood, we were relying on the bug Nokogumbo HTML5 parser has by producing the same output as Nokogiri::HTML4 parser. In turn, we were able to cleanup the resulting HTML with https://github.com/rgrove/sanitize transformers afterwards.

Now, in order to upgrade to Nokogiri 1.12 we need to come up with a way to do that before we have the parsed HTML. I'd really appreciate If someone can give me an idea how to achieve that.

Thanks.

**Comment by syakovyn** [comment]
@flavorjones, exactly! You perfectly described the issue we are facing!
